lib/iris/coords.py

```diff
@@ -2775,6 +2779,366 @@ def xml_element(self, doc):
         return cellMethod_xml_element
 
 
+class Connectivity(_DimensionalMetadata):
+    """
+    A CF-UGRID topology connectivity, describing the topological relationship
+    between two lists of dimensional elements. One or more connectivities
+    make up a CF-UGRID topology - a constituent of a CF-UGRID mesh.
+
+    See: https://ugrid-conventions.github.io/ugrid-conventions
+
+    """
+
+    def __init__(
+        self,
+        indices,
+        cf_role,
+        long_name=None,
+        var_name=None,
+        attributes=None,
+        start_index=0,
+        element_dim=1,
+    ):
+        """
+        Constructs a single connectivity.
+
+        Args:
+
+        * indices (numpy.ndarray or numpy.ma.core.MaskedArray or dask.array.Array):
+            The index values describing a topological relationship. Constructed
+            of 2 dimensions - the list of elements, and within each element:
+            the indices of the elements it relates to.
+            Use a :class:`numpy.ma.core.MaskedArray` if element lengths vary -
+            mask unused index 'slots' within each element. Use a
+            :class:`dask.array.Array` to keep indices 'lazy'.
+        * cf_role (str):
+            Denotes the topological relationship that this connectivity
+            describes. Made up of this array's elements, and the indexed
+            element within each element. E.g. ``face_node_connectivity``,
+            ``edge_face_connectivity``.
+
+        Kwargs:
+
+        * long_name (str):
+            Descriptive name of the connectivity.
+        * var_name (str):
+            The netCDF variable name for the connectivity.
+        * attributes (dict):
+            A dictionary containing other cf and user-defined attributes.
+        * start_index (int):
+            Either ``0`` or ``1``. Denotes whether :attr:`indices` uses 0- or
+            1-based indexing (allows support for Fortran and legacy NetCDF
+            files).
+        * element_dim (int):
+            Either ``0`` or ``1``. Denotes which dimension of :attr:`indices`
+            refers to the contents of individual elements (allows support
+            for fastest varying index being either first or last). E.g. for
+            ``face_node_connectivity``, where the faces are 3-sided:
+            ``indices.shape[element_dim] = 3``.
```

Comment:
In the [UGRID Conventions](https://ugrid-conventions.github.io/ugrid-conventions/), this isn't an expected attribute of a `_connectivity` variable (unlike `Connectivity`'s other parameters), but is important to include based on UGRID Conventions notes regarding `face_dimension` and `edge_dimension`.

Reply:
@trexfeathers The definition of `element_dim` is the complete **opposite** to my expectation, given the meaning of UGRID `face_dimension` and `edge_dimension` e.g.,
```shell
dimensions:
nMesh2_node = 4 ; // nNodes
nMesh2_edge = 5 ; // nEdges
nMesh2_face = 2 ; // nFaces

Two = 2 ;
Three = 3 ;

variables:
// Mesh topology
integer Mesh2 ;
Mesh2:cf_role = "mesh_topology" ;
Mesh2:long_name = "Topology data of 2D unstructured mesh" ;
Mesh2:topology_dimension = 2 ;

Mesh2:face_node_connectivity = "Mesh2_face_nodes" ;
Mesh2:face_dimension = "nMesh2_face" ;

integer Mesh2_face_nodes(nMesh2_face, Three) ;
Mesh2_face_nodes:cf_role = "face_node_connectivity" ;
Mesh2_face_nodes:long_name = "Maps every triangular face to its three corner nodes." ;
Mesh2_face_nodes:start_index = 1 ;
```
So for the above example, `face_dimension` refers to the named dimension `nMesh2_face`, which is dimension `0` of the connectivity variable `Mesh2_face_nodes`. This is the **opposite** to this doc-string explanation. Here you're saying that `element_dim` would be `1`.

... and yet later for `Connectivity.cf_role_element` you use `element` to mean the `face_dimension` dimension, not the `element_dim` dimension... I'm finding the nomenclature not quite consistent, which is confusing me.

Reply:
This is probably related to my difficulty in talking about the structure of the `indices` array and what each dimension is for. I should have just taken my lead from the UGRID conventions rather than converting to my own [difficult to explain] understanding.

Reply:
See https://github.com/SciTools/iris/pull/3968#discussion_r570120871 for an agreed solution.

Reply:
6139e563acdd0805e72a4a3e2ffa4391c3e07337